File: Smallest SubArr With Sum/Sliding Window*/slidingWindow.ts
const smallestSubWithSum = (x: number, arr: number[]) => {
  const n = arr.length;
  let left = 0,
    windowSum = 0,
    minLen = Infinity;

  for (let right = 0; right < n; right++) {
    windowSum += arr[right];

    while (windowSum > x) {
      const len = right - left + 1;
      minLen = Math.min(minLen, len);
      windowSum -= arr[left];
      left++;
    }
  }
  return minLen === Infinity ? 0 : minLen;
};

const array = [1, 4, 45, 6, 10, 19];
const num = 51;
console.log('Smallest subarray length:', smallestSubWithSum(num, array));

File: 2. Smallest SubArr With Sum/Sliding Window*/slidingWindow.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

int smallestSubWithSum(int x, const vector<int>& arr) {
  int n = arr.size();
  int left = 0, windowSum = 0, minLen = INT_MAX;

  for (int right = 0; right < n; ++right) {
    // Expand the window until windowSum <= x
    windowSum += arr[right];

    while (windowSum > x) {
      int len = right - left + 1;
      minLen = min(minLen, len);
      // Shrink the window
      windowSum -= arr[left++];
      // window = [left, right];
    }
  }
  return (minLen == INT_MAX) ? 0 : minLen;
}

int main() {
    vector<int> arr = {1, 4, 45, 6, 10, 19};
    int x = 51;

    int result = smallestSubWithSum(x, arr);
    cout << result << endl;
    return 0;
}
